jetty-core/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java

```diff
@@ -2068,6 +2070,35 @@ private ConfigurationParser(Pool.Entry<ConfigurationParser> entry)
             addCatalog(URI.create(catalogUrl.toExternalForm()));
         }
 
+        protected SAXParserFactory newSAXParserFactory()
+        {
+            // Use JVM default implementation (not the one found in the classloader from non-JVM sources)
+            SAXParserFactory factory = SAXParserFactory.newDefaultInstance();
+            // Don't allow use of XInclude to reference external entities.
+            factory.setXIncludeAware(false);
+            return factory;
+        }
+
+        @Override
+        public void setValidating(boolean validating)
+        {
+            super.setValidating(validating);
+
+            try
+            {
+                XMLReader xmlReader = getSAXParser().getXMLReader();
+
+                // disable all external entity references with Jetty's Configuration XML.
+                xmlReader.setFeature("http://xml.org/sax/features/external-general-entities", false);
+                xmlReader.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
+                xmlReader.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false);
+            }
+            catch (SAXException e)
+            {
+                LOG.warn(e.getMessage());
+            }
+        }
+
```

Comment:
Would this not work:

```suggestion
        protected SAXParserFactory newSAXParserFactory()
        {
            // Use JVM default implementation (not the one found in the classloader from non-JVM sources)
            SAXParserFactory factory = SAXParserFactory.newDefaultInstance();
            // Don't allow use of XInclude to reference external entities.
            factory.setXIncludeAware(false);

            if (isValidating())
            {
                setFeature(factory, "http://xml.org/sax/features/external-general-entities", false);
                setFeature(factory, "http://xml.org/sax/features/external-parameter-entities", false);
                setFeature(factory, "http://apache.org/xml/features/nonvalidating/load-external-dtd", false);
            }
            
            return factory;
        }

        private void setFeature(SAXParserFactory factory, String name, boolean value)
        {
            try
            {
                factory.setFeature(name, value);
            }
            catch (Exception e)
            {
                LOG.warn(e.getMessage(), e);
            }
        }
```

Reply:
Some of those are Factory features, some are XMLReader features (not all factory features translate to the created SAXParser or XMLReader, as those features can select and/or initialize the SAXParser or XMLReader you get as a result of the factory feature)
Also, the XmlParser.setValidating(boolean) method must be called for `_parser` to be initialized properly.
And the check if validating is true or not is irrelevant for XmlConfiguration.

Reply:
If you want to fix this odd initialization in XmlReader, lets do that in a different PR, for like release 12.1.x

Reply:
@joakime this is new code in this PR that is not in the correct location and looks fragile.   If the factory features are different to the parser, then wrap the the factory so it configures all parsers that it creates.    Your current code is using a side effect of a setter that known the lifecycle of a specific parser, so it is very fragile.

Reply:
There are 2 different setFeature usages.

* Factory - the setFeature calls for this are in the .newSAXParserFactory() call already.
* SAXParser - these are in the unfortunately named `.setValidating(boolean)` where the internal SAXParser is created.

Let me try a different approach, instead of wrapping SAXParserFactory, which is just as fragile.

Reply:
Different approach pushed.